Sources/CvSelectionGroupAI.cpp

```diff
@@ -669,67 +646,43 @@ CvUnit* CvSelectionGroupAI::AI_getBestGroupAttacker(const CvPlot* pPlot, bool bP
 {
 	PROFILE_FUNC();
 
-	CLLNode<IDInfo>* pUnitNode = headUnitNode();
-	const bool bIsHuman = (pUnitNode != NULL) ? GET_PLAYER(::getUnit(pUnitNode->m_data)->getOwnerINLINE()).isHuman() : true;
+	const bool bIsHuman = isHuman();
 
 	const MoveCheck::flags moveCheckFlags = MoveCheck::Attack |
 		(bPotentialEnemy ? MoveCheck::DeclareWar : MoveCheck::None) |
 		MoveCheck::CheckForBest |
 		(bAssassinate ? MoveCheck::Assassinate : MoveCheck::None) |
 		(bSuprise ? MoveCheck::Suprise : MoveCheck::None);
-			
+
 	int iBestValue = 0;
-	int iBestOdds = 0;
 	CvUnit* pBestUnit = NULL;
 
-	while (pUnitNode != NULL)
+	for (unit_iterator unitItr = beginUnits(); unitItr != endUnits(); ++unitItr)
 	{
-		CvUnit* pLoopUnit = ::getUnit(pUnitNode->m_data);
-		pUnitNode = nextUnitNode(pUnitNode);
-
+		CvUnit* pLoopUnit = *unitItr;
 		if (!pLoopUnit->isDead())
 		{
-			if ( pBestUnit != NULL && isClearlySuperior(pBestUnit, pLoopUnit, pPlot) )
+			if (pBestUnit != NULL && isClearlySuperior(pBestUnit, pLoopUnit, pPlot))
 			{
 				continue;
 			}
 
-			bool bCanAttack = false;
-			if (pLoopUnit->getDomainType() == DOMAIN_AIR)
-			{
-				bCanAttack = pLoopUnit->canAirAttack();
-			}
-			else
-			{
-				bCanAttack = pLoopUnit->canAttack();
-
-				if (bCanAttack && bNoBlitz && pLoopUnit->isBlitz() && pLoopUnit->isMadeAttack())
-				{
-					bCanAttack = false;
-				}
-			}
-
-			if (bCanAttack && (!pLoopUnit->AI_getHasAttacked()|| bSuprise))
+			if (pLoopUnit->getDomainType() == DOMAIN_AIR ? pLoopUnit->canAirAttack() : (pLoopUnit->canAttack() && !(bNoBlitz && pLoopUnit->isBlitz() && pLoopUnit->isMadeAttack())))
 			{
-				if (bForce || pLoopUnit->canMove())
+				if ((!pLoopUnit->AI_getHasAttacked() || bSuprise) && (bForce || pLoopUnit->canMove()))
 				{
 					CvUnit* pBestDefender = NULL;
 					if (bForce || pLoopUnit->canMoveInto(pPlot, moveCheckFlags, &pBestDefender))
 					{
 						PROFILE("AI_getBestGroupAttacker.RegularAttackOdds");
 
-						const int iOdds = pBestDefender? 
```

Comment:
I'm talking about here: `iOdds` is the value returned by `AI_attack*` functions, `iValue` is the same but further modified below. That makes `iBestOdds` and `iBestValue` different, so you can't just remove iBestOdds. Did I miss something here?